From e230f9e85abf1d51808e6f7bb4b3d15be3a3b3b1 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Tue, 04 Feb 2020 08:16:16 +0200
Subject: [PATCH] [SW-8228] goya2: arc: add framework to use exported symbols

This way RAM code can use ROM functions e.g. security functions from SDK
It will make RAM code footprint much smaller to fit ICCM

Change-Id: I52b6304838b73240810a4e43dec0478f37ac442f
---

diff --git a/Makefile b/Makefile
index 9b3562f..fe964fe 100644
--- a/Makefile
+++ b/Makefile
@@ -12,7 +12,7 @@
 ARCBOOT_VER := 0x0
 
 LIBDIR = sdk/lib
-ARCBOOT_CFLAGS = -Wall -W -Werror -O3
+ARCBOOT_CFLAGS = -Wall -W -Werror -O3 -g
 ARCBOOT_CFLAGS += -Iinclude
 ARCBOOT_CFLAGS += -Icommon
 ARCBOOT_CFLAGS += -I$(HL_DEV)
@@ -49,7 +49,8 @@
 
 .PHONY: all libsb libellipsys sbtool
 all: libsb libellipsys
-	@$(MAKE) arcboot_ram.bin arcboot_rom.bin
+	@$(MAKE) arcboot_rom.bin
+	@$(MAKE) arcboot_ram.bin
 
 libsb libellipsys:
 	@+$(MAKE) -C sdk $@ SB_SIGN_MODE=rsa_fixed > /dev/null
@@ -59,9 +60,10 @@
 	@$(OBJCOPY) -O binary -R .note.gnu.build-id -R .test* $< $@
 
 arcboot_ram: $(ARCBOOT_RAM_OBJECTS)
-	@$(CC) $^ $(LDFLAGS) -o $@
+	@$(CC) $^ $(LDFLAGS) -Wl,--just-symbols=exports.elf -o $@
 arcboot_rom: $(ARCBOOT_ROM_OBJECTS)
 	@$(CC) $^ $(LDFLAGS) -o $@
+	@$(OBJCOPY) --strip-all --keep-symbols=app/$@.txt $@ exports.elf
 
 %.o: %.c
 	@$(CC) $(ARCBOOT_CFLAGS) -c $< -o $@
diff --git a/app/arcboot_rom.txt b/app/arcboot_rom.txt
new file mode 100644
index 0000000..81056c6
--- /dev/null
+++ b/app/arcboot_rom.txt
@@ -0,0 +1 @@
+# put list of symbols, one per line, to be exported
